tcpdialer.go

```diff
@@ -160,11 +160,25 @@ type tcpDialer struct {
 	once sync.Once
 }
 
-const maxDialConcurrency = 1000
+var maxDialConcurrency = 1000
+
+// SetMaxDialConcurrency sets the maximum number of concurrent Dails
+// that can be performed. Setting this to 0 means unlimited.
+//
+// The initial max concurrency is set to 1000.
+//
+// WARNING: This function needs to be called before any Dials are
+// performed. Calling this function after calling any of the Dial
+// functions might not have any effect and can cause a race condition.
+func SetMaxDialConcurrency(concurrency int) {
```

Comment:
It would be better to add ability to override default dial concurrency per each dialer. Global setting for all the dialers smells not very good.

Reply:
The problem is that there is no such thing as a dialer. The only thing exposed by the api is the `Dial*` functions. Giving those an option and be backwards compatible means we have to introduce 4 new functions with options for all 4 `Dial*` functions.

What we could do is implement a `NewDialer` function that returns a `Dialer` interface that exposes a `Dial` function and some other functions to set options (such as `SetDualStack`, `SetTimeout` and `SetConcurrency`). @valyala @kirillDanshin If you guys think this is a good idea I'll update this pull request.

Reply:
@erikdubbelboer I think it's a good idea to add such dialer and expose defaultDialer.Dial* methods as a Dial* API. In such way, we can allow us to add an unlimited number of options to the dialer without having to pay for interface calls.

Reply:
Agreed with @kirillDanshin on exposing `Dialer`